### ISSUE TITLE
conftest inside of a 'nix builld' is coredumping.

### ISSUE THREAD

**Issue by nixinator** [NixOS/nixpkgs#154938]
### Describe the bug
conftest within build seems to be coredumping.
`Jan 12 21:03:07 bigNIXOS systemd-coredump[646276]: Process 646262 (conftest) of user 30016 dumped core.`

### Steps To Reproduce
1. nix build --option substitute false nixpkgs#gettext --rebuild -L
2. journalctl -f 

### Expected behavior
conftest should not be core dumping

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.81, NixOS, 21.05.20220103.88579ef (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20211006_53e4794`
 - channels(user): `"nixgl, stable-21.05.4709.88579effa7e"`
 - channels(root): `"nixos-21.11pre332093.7fad01d9d5a"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by 06kellyjac** [comment]
Using that command it built just fine and I couldn't reproduce your issue:

```
λ cat /etc/nix/registry.json | jq '.flakes[] | select(.from.id=="nixpkgs") | .to.rev' -r
78cd22c1b8604de423546cd49bfe264b786eca13
```

**Comment by nixinator** [comment]
Yep, I get rather the same problems, with the same programs.

```[542256.918863] git[2030187]: segfault at 40 ip 0000000000624e99 sp 00007fffffff5df0 error 4 in git[405000+25c000]
[542256.918883] Code: e0 00 00 00 4c 89 f6 83 c3 01 e8 d2 f4 ff ff 39 5c 24 10 0f 84 80 07 00 00 49 8b 96 60 02 00 00 48 63 c3 48 8b 12 48 8b 34 c2 <8b> 4e 40 48 89 b4 24 e0 00 00 00 8d 41 6d 39 e8 0f 8e 71 ff ff ff
[542257.617623] git[2031218]: segfault at 40 ip 0000000000624e99 sp 00007fffffff5fc0 error 4 in git[405000+25c000]
[542257.617643] Code: e0 00 00 00 4c 89 f6 83 c3 01 e8 d2 f4 ff ff 39 5c 24 10 0f 84 80 07 00 00 49 8b 96 60 02 00 00 48 63 c3 48 8b 12 48 8b 34 c2 <8b> 4e 40 48 89 b4 24 e0 00 00 00 8d 41 6d 39 e8 0f 8e 71 ff ff ff
[595516.323681] conftest[3340247]: segfault at 0 ip 00007ffff7ea4b77 sp 00007fffffffd1e8 error 4 in libc-2.33.so[7ffff7e25000+143000]
[595516.323703] Code: 00 00 00 48 39 d1 0f 82 d7 22 06 00 0f 1f 80 00 00 00 00 48 89 f8 48 83 fa 10 0f 82 fa 00 00 00 48 83 fa 20 0f 87 3f 01 00 00 <0f> 10 06 0f 10 4c 16 f0 0f 11 07 0f 11 4c 17 f0 c3 0f 1f 84 00 00
[596230.960038] conftest[3428804]: segfault at 0 ip 00007ffff7ea4b77 sp 00007fffffffc2f8 error 4 in libc-2.33.so[7ffff7e25000+143000]
[596230.960055] Code: 00 00 00 48 39 d1 0f 82 d7 22 06 00 0f 1f 80 00 00 00 00 48 89 f8 48 83 fa 10 0f 82 fa 00 00 00 48 83 fa 20 0f 87 3f 01 00 00 <0f> 10 06 0f 10 4c 16 f0 0f 11 07 0f 11 4c 17 f0 c3 0f 1f 84 00 00
[596458.712865] conftest[3560403]: segfault at 0 ip 00007ffff7ea4b77 sp 00007fffffffd538 error 4 in libc-2.33.so[7ffff7e25000+143000]
[596458.712882] Code: 00 00 00 48 39 d1 0f 82 d7 22 06 00 0f 1f 80 00 00 00 00 48 89 f8 48 83 fa 10 0f 82 fa 00 00 00 48 83 fa 20 0f 87 3f 01 00 00 <0f> 10 06 0f 10 4c 16 f0 0f 11 07 0f 11 4c 17 f0 c3 0f 1f 84 00 00
[597203.380496] conftest[3697589]: segfault at 0 ip 00007ffff7ea4b77 sp 00007fffffffc7c8 error 4 in libc-2.33.so[7ffff7e25000+143000]
[597203.380513] Code: 00 00 00 48 39 d1 0f 82 d7 22 06 00 0f 1f 80 00 00 00 00 48 89 f8 48 83 fa 10 0f 82 fa 00 00 00 48 83 fa 20 0f 87 3f 01 00 00 <0f> 10 06 0f 10 4c 16 f0 0f 11 07 0f 11 4c 17 f0 c3 0f 1f 84 00 00
[611985.006979] traps: git[2813609] general protection fault ip:4ef240 sp:7fffffff7c60 error:0 in git[405000+25c000]
[611985.341404] traps: git[2814122] general protection fault ip:4ef240 sp:7fffffff8160 error:0 in git[405000+25c000]
[611985.693819] git[2814650]: segfault at 40 ip 0000000000624809 sp 00007fffffff6f00 error 4 in git[405000+25c000]
[611985.693840] Code: e0 00 00 00 4c 89 f6 83 c3 01 e8 d2 f4 ff ff 39 5c 24 10 0f 84 80 07 00 00 49 8b 96 60 02 00 00 48 63 c3 48 8b 12 48 8b 34 c2 <8b> 4e 40 48 89 b4 24 e0 00 00 00 8d 41 6d 39 e8 0f 8e 71 ff ff ff
[611986.512956] git[2815795]: segfault at 40 ip 0000000000624809 sp 00007fffffff70c0 error 4 in git[405000+25c000]
[611986.512975] Code: e0 00 00 00 4c 89 f6 83 c3 01 e8 d2 f4 ff ff 39 5c 24 10 0f 84 80 07 00 00 49 8b 96 60 02 00 00 48 63 c3 48 8b 12 48 8b 34 c2 <8b> 4e 40 48 89 b4 24 e0 00 00 00 8d 41 6d 39 e8 0f 8e 71 ff ff ff
[612894.175188] traps: git[3401592] general protection fault ip:4ef360 sp:7fffffff6b50 error:0 in git[405000+25c000]
[612894.484980] traps: git[3402098] general protection fault ip:4ef360 sp:7fffffff7060 error:0 in git[405000+25c000]
[612894.980031] git[3402853]: segfault at 40 ip 0000000000624e99 sp 00007fffffff5df0 error 4 in git[405000+25c000]
[612894.980051] Code: e0 00 00 00 4c 89 f6 83 c3 01 e8 d2 f4 ff ff 39 5c 24 10 0f 84 80 07 00 00 49 8b 96 60 02 00 00 48 63 c3 48 8b 12 48 8b 34 c2 <8b> 4e 40 48 89 b4 24 e0 00 00 00 8d 41 6d 39 e8 0f 8e 71 ff ff ff
[612895.829408] git[3403649]: segfault at 40 ip 0000000000624e99 sp 00007fffffff5fc0 error 4 in git[405000+25c000]
[612895.829429] Code: e0 00 00 00 4c 89 f6 83 c3 01 e8 d2 f4 ff ff 39 5c 24 10 0f 84 80 07 00 00 49 8b 96 60 02 00 00 48 63 c3 48 8b 12 48 8b 34 c2 <8b> 4e 40 48 89 b4 24 e0 00 00 00 8d 41 6d 39 e8 0f 8e 71 ff ff ff
[613985.160003] testp[3888605]: segfault at 0 ip 00000000004024cc sp 00007fffffffb920 error 6 in testp[402000+9a000]
[613985.160024] Code: 01 74 0b 83 7d fc 02 74 2c e9 24 01 00 00 83 7d f8 00 74 0c 48 8d 3d a4 9b 09 00 e8 ce fb ff ff 48 c7 c0 78 15 58 00 48 8b 00 <c7> 00 00 00 00 00 e9 1a 01 00 00 83 7d f8 00 74 0c 48 8d 3d 94 9b
[614025.801631] testnoshared[3891334]: segfault at 0 ip 00000000004024cc sp 00007fffffffb910 error 6 in testnoshared[402000+9a000]
[614025.801651] Code: 01 74 0b 83 7d fc 02 74 2c e9 24 01 00 00 83 7d f8 00 74 0c 48 8d 3d a4 9b 09 00 e8 ce fb ff ff 48 c7 c0 78 15 58 00 48 8b 00 <c7> 00 00 00 00 00 e9 1a 01 00 00 83 7d f8 00 74 0c 48 8d 3d 94 9b
[618877.007151] traps: llc[4071863] general protection fault ip:7ffff266b71f sp:7fffffffd770 error:0 in libLLVM-13.so[7ffff201d000+36e3000]
```

**Comment by ius** [comment]
The `nix` segfault looks like a null pointer dereference and might be unrelated to the rest..

**Comment by cole-h** [comment]
TL;DR: Just a noisy failure of a feature test, which shouldn't impact anything.

----

I think I've found the source of this segfault. I didn't realize that `conftest` is just the name of the binary for all the `autoconf/make/gen/whatever` functionality tests (e.g. the `checking for wchar_t... yes` you see flying by), so it took a bit of digging until I realized this.

I decided to just clone `gettext` from source and try to build it without the Nix sandbox to make sure it wasn't just an issue with that. I needed to use an older version of `autoconf` and `automake`, because the v0.21 tag only supports up to `autoconf 2.69` and `automake 1.16` (21.05+ have newer versions), but after that I was able to get a more-informative segfault that hinted at its origin:

```
Jan 21 11:30:27 scadrial systemd-coredump[988795]: [🡕] Process 988793 (conftest) of user 1000 dumped core.
                                                   
                                                   Found module /home/vin/workspace/vcs/gettext/libtextstyle/conftest without build-id.
                                                   Found module linux-vdso.so.1 with build-id: 4d8933d129df4d3bca615f305023b1a171cd9795
                                                   Found module ld-linux-x86-64.so.2 with build-id: 38035ba01548a26f962b62aacd1f457428c44b60
                                                   Found module libc.so.6 with build-id: fe5202e764d559678d6a40b026c77838ba5dda44
                                                   Stack trace of thread 988793:
                                                   #0  0x00007f15ce4cdbaa raise (libc.so.6 + 0x3bbaa)
                                                   #1  0x00007f15ce4b8523 abort (libc.so.6 + 0x26523)
                                                   #2  0x00007f15ce50e2e8 __libc_message (libc.so.6 + 0x7c2e8)
                                                   #3  0x00007f15ce50e312 __libc_fatal (libc.so.6 + 0x7c312)
                                                   #4  0x00007f15ce4fcd2d __vfprintf_internal (libc.so.6 + 0x6ad2d)
                                                   #5  0x00007f15ce5072be __vsprintf_internal (libc.so.6 + 0x752be)
                                                   #6  0x00007f15ce5999c9 __sprintf_chk (libc.so.6 + 0x1079c9)
                                                   #7  0x00000000004010b0 n/a (/home/vin/workspace/vcs/gettext/libtextstyle/conftest + 0x10b0)
```

So I then focused my efforts inside the `libtextstyle` subdir and rigged up a "watcher" that would kill the configure process as soon as possible after it logged a segfault using the following:

```
$ journalctl -ft systemd-coredump --since @(date +%s) | awk '/dumped core/ { system("pkill -f \"^/nix/store/ynpihpad0201sz78m8lnl7m0x6r1ks1b-bash-5.1-p8/bin/bash ./configure\"") }'
```

which gave even more hints at the issue -- the last few lines of logging that were output after it was terminated were:

```
checking whether printf supports the 'F' directive... yes
checking whether printf supports the 'n' directive... no
checking whether printf supports the 'ls' directive... yes
[...]
checking whether printf survives out-of-memory conditions... Terminated
```

Then, I searched inside the `config.log` for `core dump` and related strings until I found that it happened while it was `checking whether printf supports the 'n' directive`. It prints the file that failed in the log, so I copied the contents out and compiled it myself (for the curious, I've uploaded it as a gist here: https://gist.github.com/cole-h/74fcf00541f68f4da1d405593ddfb7c5). After running it outside of the `nix develop` shell, I was unable to reproduce (and yet I could inside of it), so that led me to look at my env vars...

I tried adding pretty much every `NIX_*` env var to my outside-`nix develop` invocation until I found a combination that reproduced the issue:

```
$ NIX_HARDENING_ENABLE='fortify stackprotector pic strictoverflow relro bindnow' NIX_CC_WRAPPER_TARGET_HOST_x86_64_unknown_linux_gnu=1 gcc -o conftest -g -O2 conftest.c
$ ./conftest
*** %n in writable segment detected ***
fish: Job 1, './conftest' terminated by signal SIGABRT (Abort)
```

(Yes, both the `NIX_HARDENING_ENABLE` and `NIX_CC_WRAPPER_TARGET_HOST` env vars are necessary to reproduce.)

Next, I went to ye olde search engine and looked around for where `*** %n in writable segment detected ***` came from. This led me to the `gnulib` mailing list: https://lists.gnu.org/archive/html/bug-gnulib/2009-02/msg00106.html. So it looks like the `fortify` hardening is what causes this (though I still don't know why the `NIX_CC_WRAPPER_TARGET_HOST` env var is also necessary to reproduce...). However, this appears to be OK; quoting Bruno Habile from that email: "This is fine: it explains why on this glibc system, the snprintf is not fully ISO C99 compliant."

So this just looks like the feature test is failing (though in a somewhat noisy way...), and all-in-all I agree with Bruno that this is "fine".

---

Since I don't think this is an problem with Nixpkgs, should this issue be closed?

**Comment by nixinator** [comment]
Amazing debug, thanks for taking the time to get to the bottom of this.  However maybe my idea of fine and the developers idea of fine maybe different. I wonder if it's down to a memory layout to recreate it, i.e. when you add  NIX_CC_WRAPPER_TARGET_HOST , the memory layout is slightly different that triggers it? 

Seems minor, but annoying especially add noise to the logs... 

https://tenor.com/view/everything-is-fine-dont-look-up-gif-24325668

**Comment by grahamc** [comment]
From the [linked issue](https://lists.gnu.org/archive/html/bug-gnulib/2009-02/msg00106.html), it is running this test to perform feature detection. As far as I can tell the segfault means the test failed. The linked issue doesn't even suggest that making it stop segfaulting is a valid goal, but that the issue is in how glibc handles and displays the segfault.

I don't think there is anything we can do here, since we're not going to disable `fortify` hardening. Probably the right next step if you wanted to dig in to this is to engage further with the gnulib and glibc communities, and try to find an alternative way to perform the detection.

**Comment by nixinator** [comment]
All of you, thanks for the excellent detective work. I will just ignore if for now. Why is git doing it as well, or is that just conftest again, but somehow mistakenly identifying the process is git?

**Comment by nixinator** [comment]
shall i open a new issue on git core dumping in the build sandbox, or reopen this one? Not sure.

**Comment by cole-h** [comment]
Please open a new one, yes.

**Comment by nixinator** [comment]
> Please open a new one, yes.

I'm trying to pin point what builds are doing this. 

https://github.com/NixOS/nixpkgs/issues/157757
